Implement Auditing for Cases

murilovmachado opened the pull request:
Jira: https://bugs.benetech.org/browse/CHI-129

This PR:
- Creates `CaseAudit`:
  - Model
  - Migration
- Adds `twilioWorkerId` to `Case`
- Create Hooks:
  - on `Case.afterUpdate`: record when updating a case
  - on `Contact.afterUpdate`: record when connecting a contact to a case

nickhurlburt commented on models/case-audit.js:
```diff
@@ -0,0 +1,17 @@
+     * But sequelize was NOT inserting 'caseId' when CREATING a CaseAudit.
+     */
+    caseId: DataTypes.INTEGER,
+  });
```
Would we want to capture a revision id?  Would we want to have the timestamp of the change at the top level here, or is having it in `newValue.dataValues.updatedAt` sufficient?

murilovmachado replied:
`CaseAudit` will have by default `createdAt` and `updatedAt` columns as the other tables.

nickhurlburt replied:
Okay.  Seems like that should work fine.

murilovmachado commented on models/case.js:
```diff
@@ -3,9 +3,37 @@ module.exports = (sequelize, DataTypes) => {
+    const previousValue = {
+      // eslint-disable-next-line no-underscore-dangle
+      ...caseInstance._previousDataValues,
+      contacts: contactsId,
```
Right, an update on `Case` itself wouldn't change the associated contacts, since this information resides in the `Contact` table. You're also right when you say a contact change would be tracked by the hook in the `Contact` model.

murilovmachado commented on models/contact.js:
```diff
@@ -1,3 +1,66 @@
+  caseFromDB,
+  transaction,
+) => {
+  if (!caseFromDB) return;
```
Exactly, the first time a contact is created it doesn't have any associated case yet (`caseId` is `null`). So when it gets connected, there's no previous case to audit.

murilovmachado commented on models/contact.js:
```diff
@@ -1,3 +1,66 @@
+
+const getPreviousAndNewCases = async (contactInstance, transaction) => {
+  const { Case } = contactInstance.sequelize.models;
+  const previousCaseId = contactInstance._previousDataValues.caseId;
```
I didn't know about `previous`. I'll investigate to see if it handles everything we need.

murilovmachado replied:
so, `previous(key)` will return the previous value for that `key`, and `previous()` without argument will return the previous values that have changed. There's no direct substitution for `_previousDataValues`, but we can easily achieve that with `{...contactInstance.dataValues, ...contactInstance.previous()}`. I'll make the change, just in case they endup making `_previousDataValues` unaccessible in the future.

nickhurlburt replied:
That sounds good to me, but one thought:  could doing `{...contactInstance.dataValues, ...contactInstance.previous()}` mean that we end up including a value that didn't previously exist?  Like, let's say we added `"foo": "bar"` and it's now in the `dataValues`.  Would `previous()` return `"foo": undefined` and then it would be omitted?

murilovmachado replied:
I actually don't know how `previous()` would behave in this situation, and it may lead to this issue of including a value that shouldn't exist. In practice, this is currently unlikely to happen because `Case` will have all properties set in its creation. But now I feel our code could be very vulnerable in the future because of this. 

Let me try to test this scenario and see what happens in this situation. I'll get back with an answer.

murilovmachado replied:
I've investigated it a little bit more, and it behave correclty: `previous()` brought `newProp: null`. I believe it's safe to go on with this approach.

murilovmachado commented on models/case.js:
```diff
@@ -3,9 +3,37 @@ module.exports = (sequelize, DataTypes) => {
+    const contactsId = contacts.map(contact => contact.dataValues.id);
+    const previousValue = {
+      // eslint-disable-next-line no-underscore-dangle
+      ...caseInstance._previousDataValues,
```
`_previousDataValues` capture all values. `.previous()` I don't know, I'll investigate it.

murilovmachado commented on models/case.js:
```diff
@@ -3,9 +3,37 @@ module.exports = (sequelize, DataTypes) => {
+    };
+    const { CaseAudit } = caseInstance.sequelize.models;
+    await CaseAudit.create(caseAuditRecord, {
+      transaction: options.transaction,
```
Right now, yes. It's all in a single transaction. But, I believe we can have 2 separate transactions:
1) One for the `Case` update
2) One for handling the auditing part

Would that be better?

nickhurlburt replied:
From a data integrity point of view, I definitely like that it's all one transaction.  I don't like situations where the `Case` update can happen but the audit record could fail.  I don't know what the performance impact would be, and I assume for now that it's not serious.

murilovmachado commented:
@nickhurlburt This PR focus on recording **every change** to `Case` so we can audit them later on. It definitely records all the information needed to answer these questions:

> - Which counselor changed the status from 'Open' to 'Closed'?
> - Who has edited the 'Case summary', when, and how?
> - Who originally opened the case?

For this last question, we will need to add hooks on `Perpetrator` and `Household` like I did for `Contact`:
- Has anyone disassociated or deleted other objects that were connected to the case, like perpetrators or household information?

For consistency, I think it would be good to put a hook on `Case.afterCreate` as well because, now, we're only creating an audit record on `Case.afterUpdate`. (Although we're not currently missing any info, because we're storing `previousValue`, so the original value will be there)

But what this PR does not intend to answer is: how to retrieve this audit data or how to present this information. It's just about making sure we're recording every change.

nickhurlburt commented:
> For consistency, I think it would be good to put a hook on Case.afterCreate as well because, now, we're only creating an audit record on Case.afterUpdate. (Although we're not currently missing any info, because we're storing previousValue, so the original value will be there)

Makes sense to me.

> But what this PR does not intend to answer is: how to retrieve this audit data or how to present this information. It's just about making sure we're recording every change.

Yes, that sounds good and I understand.  I'm just double-checking that we're in sync about the overall goal, and that you have that goal in mind even if this is only a part of it.  As far as I can tell, this will address that goal just fine but it's always worth restating!

murilovmachado commented:
@nickhurlburt I've addressed all the issues discussed in the PR. Please take another look when you can. I also think I should create tests for these hooks as well.

nickhurlburt commented on models/case.js:
```diff
@@ -1,11 +1,58 @@
+    const contactIds = await getContactIds(caseInstance);
+    const previousValue = {
+      ...caseInstance.dataValues,
+      ...caseInstance.previous(),
```
What was the outcome on how this would handle new keys that were added?  Did that not end up being an issue?

murilovmachado replied:
I've investigated and `previous()` would bring any  `newKey` as `null`, so it won't be an issue.
